cpp/src/quantiles/quantiles.cu

```diff
@@ -20,60 +20,104 @@
+      cudaMemcpy(&hv[0], d_res, sizeof(T), cudaMemcpyDeviceToHost);//TODO: async with streams?
+      result = static_cast<RetT>( hv[0] );
+      return GDF_SUCCESS;
+    }
 
-    FctrType midpoint{[](VType y0, VType y1, double x){
-        return static_cast<RetT>(static_cast<double>(y0 + y1)/2.0);
-      }};
+    if( q <= 0.0 && flag_sorted )
+    {
+      T* d_res = thrust::min_element(rmm::exec_policy(stream)->on(stream), dv, dv+n);
+      cudaMemcpy(&hv[0], d_res, sizeof(T), cudaMemcpyDeviceToHost);//TODO: async with streams?
+      result = static_cast<RetT>( hv[0] );
+      return GDF_SUCCESS;
+    }
 
-    FctrType nearest{[](VType y0, VType y1, double x){
-        return static_cast<RetT>(x < 0.5 ? y0 : y1);
-      }};
+    if( n < 2 )
+    {
+      cudaMemcpy(&hv[0], dv, sizeof(T), cudaMemcpyDeviceToHost);//TODO: async with streams?
+      result = static_cast<RetT>( hv[0] );
+      return GDF_SUCCESS;
+    }
 
-    FctrType lowest{[](VType y0, VType y1, double x){
-        return static_cast<RetT>(y0);
-      }};
+    // sort if the input is not sorted.
+    if( !flag_sorted ){
+      thrust::sort(rmm::exec_policy(stream)->on(stream), dv, dv+n);
+    }
 
-    FctrType highest{[](VType y0, VType y1, double x){
-        return static_cast<RetT>(y1);
-      }};
-  
-  
-    std::vector<std::string> methods{"lin_interp", "midpoint", "nearest", "lowest", "highest"};
-    size_t n_methods = methods.size();
-    std::vector<FctrType> vf{lin_interp, midpoint, nearest, lowest, highest};
-  
-    std::vector<double> qvals{0.0, 0.25, 0.33, 0.5, 1.0};
+    QuantiledIndex qi = find_quantile_index(n, q);
 
-  
-    assert( n_methods == methods.size() );
-  
-    for(auto q: qvals)
-      {
-        VType res = quantile_approx(d_in.data().get(), d_in.size(), q);
-        std::cout<<"q: "<<q<<"; exact res: "<<res<<"\n";
-        for(auto i = 0;i<n_methods;++i)
-          {
-            RetT rt = quantile_exact(d_in.data().get(), d_in.size(), q, vf[i]);
-            std::cout<<"q: "<<q<<"; method: "<<methods[i]<<"; rt: "<<rt<<"\n";
-          }
-      }
-  }
+    switch( interpolation )
+    {
+    case GDF_QUANT_LINEAR:
+      cudaMemcpy(&hv[0], dv+qi.lower_bound, sizeof(T), cudaMemcpyDeviceToHost); //TODO: async with streams
+      cudaMemcpy(&hv[1], dv+qi.upper_bound, sizeof(T), cudaMemcpyDeviceToHost); //TODO: async with streams
+      //TODO: safe operation to avoid overflow/underflow
+      result = static_cast<RetT>(static_cast<RetT>(hv[0]) + qi.fraction*static_cast<RetT>(hv[1]-hv[0]));
+      break;
+    case GDF_QUANT_MIDPOINT:
+      cudaMemcpy(&hv[0], dv+qi.lower_bound, sizeof(T), cudaMemcpyDeviceToHost); //TODO: async with streams
+      cudaMemcpy(&hv[1], dv+qi.upper_bound, sizeof(T), cudaMemcpyDeviceToHost); //TODO: async with streams
+      //TODO: safe operation to avoid overflow/underflow
+      result = static_cast<RetT>(static_cast<RetT>(hv[0] + hv[1])/2.0);

```

Comment:
I think this would be less likely to overflow if you formulate it as:
```
result = static_cast<RetT>(static_cast<RetT>(hv[0])/2.0 + static_cast<RetT>(hv[1])/2.0);
```

I.e., just distribute the division by `2.0`.

Reply:
The implementation is moved into header file.
https://github.com/rapidsai/cudf/blob/db1cbb9ca34aadce0323e8dc7c19ca63f67f126a/cpp/src/quantiles/quantiles.hpp#L45-L80

Also `cudf::interpolate::midpoint` is tested in gtests.
However, `cudf::interpolate::linear` is not tested, it should be resolved by  Issue #1417